mempool/llrb_mempool.go

```diff
@@ -0,0 +1,217 @@
+package mempool
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	tmmath "github.com/tendermint/tendermint/libs/math"
+	"github.com/tendermint/tendermint/types"
+
+	abcix "github.com/tendermint/tendermint/abcix/types"
+	cfg "github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/libs/llrb"
+	"github.com/tendermint/tendermint/proxy"
+)
+
+//--------------------------------------------------------------------------------
+
+// lElement is used to insert, remove, or recheck transactions
+type lElement struct {
+	nodeKey llrb.NodeKey
+	tx      *mempoolTx
+}
+
+//--------------------------------------------------------------------------------
+
+// LlrbMempool is an ordered in-memory pool for transactions before they are
+// proposed in a consensus round. Transaction validity is checked using the
+// CheckTx abci message before the transaction is added to the pool. The
+// mempool uses a left-leaning red-black tree structure for storing transactions that can
+// be efficiently accessed by multiple concurrent readers.
+type LlrbMempool struct {
+	txs llrb.LLRB // left-leaning red-black tree of good txs
+
+	// Map for quick access to txs to record sender in CheckTx.
+	// txsMap: txKey -> lElement
+	txsMap sync.Map
+
+	// Track whether we're rechecking txs.
+	// These are not protected by a mutex and are expected to be mutated in
+	// serial (ie. by abci responses which are called in serial).
+	recheckCursor *lElement // next expected response
+	recheckEnd    *lElement // re-checking stops here
+}
+
+// NewLlrbMempool returns a new mempool with the given configuration and connection to an application.
+func NewLlrbMempool(
+	config *cfg.MempoolConfig,
```

Comment:
take a look at the naming conventions, especially the last paragraph on naming acronyms https://stackoverflow.com/a/39830270